js/pro/gemini.js

```diff
@@ -0,0 +1,664 @@
+        //         type: 'l2_updates',
+        //         symbol: 'BTCUSD',
+        //         changes: [
+        //             [ 'buy', '22252.37', '0.02' ],
+        //             [ 'buy', '22251.61', '0.04' ],
+        //             [ 'buy', '22251.60', '0.04' ],
+        //             // some asks as well
+        //         ],
+        //         trades: [
+        //             { type: 'trade', symbol: 'BTCUSD', event_id: 122258166738, timestamp: 1655330221424, price: '22269.14', quantity: '0.00004473', side: 'buy' },
+        //             { type: 'trade', symbol: 'BTCUSD', event_id: 122258141090, timestamp: 1655330213216, price: '22250.00', quantity: '0.00704098', side: 'buy' },
+        //             { type: 'trade', symbol: 'BTCUSD', event_id: 122258118291, timestamp: 1655330206753, price: '22250.00', quantity: '0.03', side: 'buy' },
+        //         ],
+        //         auction_events: [
+        //             {
+        //                 type: 'auction_result',
+        //                 symbol: 'BTCUSD',
+        //                 time_ms: 1655323200000,
+        //                 result: 'failure',
+        //                 highest_bid_price: '21590.88',
+        //                 lowest_ask_price: '21602.30',
+        //                 collar_price: '21634.73'
+        //             },
+        //             {
+        //                 type: 'auction_indicative',
+        //                 symbol: 'BTCUSD',
+        //                 time_ms: 1655323185000,
+        //                 result: 'failure',
+        //                 highest_bid_price: '21661.90',
+        //                 lowest_ask_price: '21663.79',
+        //                 collar_price: '21662.845'
+        //             },
+        //         ]
+        //     }
+        //
+        this.handleOrderBook (client, message);
+        this.handleTrades (client, message);
+    }
+
+    async watchOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        /**
+         * @method
+         * @name gemini#fetchOrders
+         * @description watches information on multiple orders made by the user
+         * @see https://docs.gemini.com/websocket-api/#order-events
+         * @param {string|undefined} symbol unified market symbol of the market orders were made in
+         * @param {int|undefined} since the earliest time in ms to fetch orders for
+         * @param {int|undefined} limit the maximum number of  orde structures to retrieve
+         * @param {object} params extra parameters specific to the gemini api endpoint
+         * @returns {[object]} a list of [order structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-structure}
+         */
+        const url = this.urls['api']['ws'] + '/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked';
+        await this.loadMarkets ();
```

Comment:
@pcriadoperez unfortunately even after the market's merge, I still get an error here:
![image](https://user-images.githubusercontent.com/43336371/208655584-6530fd30-d1d6-4e0c-95a6-a4c559c931d4.png)

and does not seem related to the markets because they're correctly loaded:
![image](https://user-images.githubusercontent.com/43336371/208655647-dab228b4-26bb-47f9-8587-fddb58f45104.png)


Can you please check this out?

Reply:
Good catch! I found the issue was that we weren't including headers in python websockets. I opened a seperate PR to address this issue: #16203

Reply:
@pcriadoperez Unfortunately, even with https://github.com/ccxt/ccxt/pull/16203 merged I still get an error here. Weird thing is the error is different depending on the language:

Javascript:
```
CCXT v2.4.91
gemini.watchOrders ()
2022-12-31T15:37:51.286Z connecting to wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked
2022-12-31T15:37:51.691Z onError Unexpected server response: 400
2022-12-31T15:37:51.706Z onClose CloseEvent {
  [Symbol(kTarget)]: WebSocket {
    _events: [Object: null prototype] {
      open: [Function],
      message: [Function],
      error: [Function],
      close: [Function],
      ping: [Function: bound onPing],
      pong: [Function: bound onPong],
      upgrade: [Function: bound onUpgrade]
    },
    _eventsCount: 7,
```
```Python
p gemini watchOrders --verbose 
Python v3.10.8
CCXT v2.4.91
gemini.watchOrders()
2022-12-31T15:39:02.855Z connecting to wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked with timeout 10000 ms
2022-12-31T15:39:03.252Z NetworkError 400, message='Invalid response status', url=URL('wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked')
2022-12-31T15:39:03.252Z on_error 400, message='Invalid response status', url=URL('wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked')
Traceback (most recent call last):
  File "/Users/cjg/Git/ccxt7/ccxt/examples/py/cli.py", line 223, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/cjg/Git/ccxt7/ccxt/examples/py/cli.py", line 202, in main
    result = await method(*args)
  File "/Users/cjg/Git/ccxt7/ccxt/python/ccxt/pro/gemini.py", line 379, in watch_orders
    orders = await self.watch(url, messageHash, None, messageHash)
ccxt.base.errors.NetworkError: 400, message='Invalid response status', url=URL('wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked')
```

Reply:
strange, for me it's working for both js and python. Do you have any more info on the error to try and reproduce it?

Reply:
@pcriadoperez yeah I have no idea, I tried several locations (uk, eua, etc) and got the exact same error

Reply:
@pcriadoperez Great finding! 🙂 
Yeah the option seems a good idea, I will only change it slightly to be less dynamic. (basically `this[method]()` should be avoided IMO because it's hard to translate to typed languages)